### PR TITLE
BatchNorm guide

### PR DESCRIPTION
**Pull request by cgarciae**
# What does this PR do?

WIP

### CODE REVIEW

**Comment by zaxtax** [comment]
I think this is good, but it could help if the code could be fully copy-pasted. So maybe including all relevant import statements.

I wouldn't use the word "just" since even though it's a small change the introduction of the `train` flag is confusing and surprising. As you say, `BatchNorm` is often the first model most users encounter that behaves much more differently at train vs test time. I think that's worth dwelling on.

Minor:
"BatchNorm add an" -> "BatchNorm adds an"

**Comment by cgarciae** [comment]
Hey @zaxtax, thanks for the feedback! What did you mean by:

> I think this is good, but it could help if the code could be fully copy-pasted.

Like a "copy to clipboard" widget?

**Comment by zaxtax** [comment]
Oh I meant the example was completely self-contained and could be run if it
were cut-paste from the website directly into a notebook cell.

**Comment by cgarciae** [comment]
Makes sense! Once the content for this guide is finalized the idea is to copy this into a notebook and add an "open in colab" link.

**Review by levskaya — APPROVED** [review]
This is really awesome, I don't have any useful critiques at the moment. lgtm!!!!

File: docs/guides/batch_norm.rst [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+  import jax
+  import optax
+  from typing import Any
+  from flax.core import FrozenDict
+
+Defining the model
+******************
+
+``BatchNorm`` is a Module that has different runtime behavior between training and
+inference. In other frameworks this behavior specified via mutable state, however
```
Comment (marcvanzee) [inline docs/guides/batch_norm.rst:21]: Could you maybe add a link pointing to another framework and how it uses mutable state? It is a bit unclear to me now.

File: docs/guides/batch_norm.rst [hunk 2]
```diff
@@ -0,0 +1,294 @@
+Using BatchNorm
```
Comment (marcvanzee) [inline docs/guides/batch_norm.rst:1]:
```suggestion
Using Batch Normalization
```

File: docs/guides/batch_norm.rst [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+===============
+
+In this guide we will go through the details of using ``BatchNorm`` in models,
```
Comment (marcvanzee) [inline docs/guides/batch_norm.rst:4]: Maybe also add one sentence what batch normalization is, and link to a resource with more information?

File: docs/guides/batch_norm.rst [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+  class MLP(nn.Module):
+    @nn.compact
+    def __call__(self, x, train: bool): #!
+      x = nn.Dense(features=4)(x)
+      x = nn.BatchNorm(use_running_average=not train)(x) #!
+      x = nn.relu(x)
+      x = nn.Dense(features=1)(x)
+      return x
+
+Once the model is created initializing is very similar, the additional
```
Comment (marcvanzee) [inline docs/guides/batch_norm.rst:50]: This sentence is a bit confusing. What do you want to say with it? It is very similar to what? Maybe you should just say that if you initialize the model with `train=True`, then batch norm will compute the batch statistics, which it stores in a mutable collection, and that you explain this in more detail in the next section.

File: docs/guides/batch_norm.rst [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+``train`` argument is passed to ``init`` to get back the ``variables`` structure.
+
+The ``batch_stats`` collection
+******************************
+
+Apart from the ``params`` collection, ``BatchNorm``
```
Comment (marcvanzee) [inline docs/guides/batch_norm.rst:56]: It may be good to link to our [Variable collections documentation](https://flax.readthedocs.io/en/latest/api_reference/flax.linen.html#module-flax.core.variables) for people unfamiliar with this term.

File: docs/guides/batch_norm.rst [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+  mlp = MLP()
+  x = jnp.ones((1, 3))
+  variables = mlp.init(jax.random.PRNGKey(0), x, train=False) #!
+  params = variables['params']
+  batch_stats = variables['batch_stats'] #!
+
+  jax.tree_util.tree_map(jnp.shape, variables)
+
+
+``BatchNorm`` adds a total of 4 variables: ``mean`` and ``var`` that live in the
```
Comment (marcvanzee) [inline docs/guides/batch_norm.rst:83]: `mean` and `var` are only added if `use_running_average==False`.

File: docs/guides/batch_norm.rst [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+
+When using ``apply`` to run your model a couple of things must
+be taken into consideration:
+
+- ``batch_stats`` must be passed as an input variable.
+- The ``train`` argument must be defined.
+- During training the ``batch_stats`` collection to be marked as
+  mutable by setting ``mutable=['batch_stats']``.
+- When there are mutable collection, updates to these are returned as a
+  second output. The updated ``batch_stats`` must be extracted from here.
```
Comment (marcvanzee) [inline docs/guides/batch_norm.rst:149]:
```suggestion
When using ``apply`` to run your model with ``train==True``
(i.e., ``use_running_average==False` in the call to ``BatchNorm``),
a couple of things must be taken into consideration:

- ``batch_stats`` must be passed as an input variable.
- The ``batch_stats`` collection to be marked as
  mutable by setting ``mutable=['batch_stats']``.
- The mutated variables are returned as a second output.
  The updated ``batch_stats`` must be extracted from here.
```

File: docs/guides/batch_norm.rst [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+    {'params': params, 'batch_stats': batch_stats}, #!
+    x,
+    train=True, mutable=['batch_stats'] #!
+  )
+  batch_stats = updates['batch_stats'] #!
+
+Training and Evaluation
+***********************
+
+To actually integrate this model into a training loop that uses
```
Comment (marcvanzee) [inline docs/guides/batch_norm.rst:174]: Can you also explain *why* this is necessary?

Comment (marcvanzee) [inline docs/guides/batch_norm.rst:174]:
```suggestion
To integrate this model into a training loop that uses
```

File: docs/guides/batch_norm.rst [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+    batch_stats: Any  #!
+
+  state = TrainState.create( #!
+    apply_fn=mlp.apply,
+    params=params,
+    batch_stats=batch_stats, #!
+    tx=optax.adam(1e-3),
+  )
+
+Finally the ``train_step`` must be updated to reflect these changes, the main
```
Comment (marcvanzee) [inline docs/guides/batch_norm.rst:207]:
```suggestion
Finally the ``train_step`` function must be updated to reflect these changes, the main
```

File: docs/guides/batch_norm.rst [hunk 10]
```diff
@@ -0,0 +1,294 @@
...
+    state = state.apply_gradients(grads=grads)
+    state = state.replace(batch_stats=updates['batch_stats']) #!
+    metrics = {
+      'loss': loss,
+        'accuracy': jnp.mean(jnp.argmax(logits, -1) == batch['label']),
+    }
+    return state, metrics
+
+The ``eval_step`` is much simpler, since ``batch_stats`` is not mutable no
+updates need to be propagated. Only difference is that ``batch_stats`` must be
```
Comment (marcvanzee) [inline docs/guides/batch_norm.rst:260]:
```suggestion
updates need to be propagated. The only difference is that ``batch_stats`` must be
```

File: docs/guides/batch_norm.rst (hunk 1 above)
Comment (cgarciae) [reply docs/guides/batch_norm.rst:21]: Added 2 links, and some additional comments.

File: docs/guides/batch_norm.rst (hunk 6 above)
Comment (cgarciae) [reply docs/guides/batch_norm.rst:83]: I think they are always added: 
https://github.com/google/flax/blob/722838ce11e6b0b307dc6cdf44b4fa3175ead3a9/flax/linen/normalization.py#L248-L253

File: docs/guides/batch_norm.rst (hunk 8 above)
Comment (cgarciae) [reply docs/guides/batch_norm.rst]: Added explenation

File: docs/guides/batch_norm.rst (hunk 4 above)
Comment (cgarciae) [reply docs/guides/batch_norm.rst]: Clarified sentence.

File: docs/guides/batch_norm.rst (hunk 3 above)
Comment (cgarciae) [reply docs/guides/batch_norm.rst]: Added explanation

**Comment by cgarciae** [comment]
Hey @marcvanzee! Is the ready to go? 🚀

**Comment by marcvanzee** [comment]
Sure, I thought @8bitmp3 might want to have a look, but I think he has plenty of other tasks, and given that I already looked at it in quite some detail I think we can merge it.

**Comment by marcvanzee** [comment]
@8bitmp3 if you still have comments, feel free to create a new PR with suggested modifications.

**Review by 8bitmp3 — APPROVED** [review]
Reviewed and updated. Thank you 👍
